### DISCUSSION TITLE
Block Design Idea

### DISCUSSION THREAD

**Discussion by meganindya** [sugarlabs/musicblocks-v4#25] [Ideas]
The dynamicity of stacking connections has not been addressed very much in the block designs of Music Blocks. Certain blocks (logically) shouldn't connect to the end of certain blocks (or connect only to the end of certain blocks). Certain blocks are terminal blocks. In Music Blocks, spikes are used to express that, but there are inconsistencies here and there.

Another thing is that arguments to blocks are arranged vertically. As blocks are connected, vertical spacer blocks need to be used so that blocks don't overlap. These tend to eat up vertical screen real estate very quickly. With only a few blocks stacked, it tends to exceed the screen.

Here is an idea taking these into consideration.

### Flow Blocks

![blocks_flow](https://user-images.githubusercontent.com/33328728/107845629-e4a5e780-6e02-11eb-9348-c1b0194b2ff5.png)

- Connectors are circular rather than wedges. Think of these as magnets. The white ones attach to any connectors. The blue ones are similar; blue ones go on top of the white ones as they connect. Other color options can describe pairs of selective connectivity. Notice the 2nd block on the top-left &mdash; instead of a connector it has a protrusion of the body. No blocks attach to the end of those &mdash; such blocks could be terminal blocks in top-level clamps.

- Another thing that's new are that arguments, instead of going on the sides arranged vertically, they now go within the body of the parent blocks arranged horizontally. The labels go above them. For a lot of binary operations, I believe, this is easier to follow. And this design saves a lot of space, vertically, while expanding in limited size horizontally.

### Clamp Blocks

![blocks_clamp](https://user-images.githubusercontent.com/33328728/107846347-0bffb300-6e09-11eb-815f-14ba93a112f2.png)

These are similar to the flow blocks apart from the connection at the end of the clamp, which again may be selective. Notice the top-level blue block that doesn't attach to any other block on the outside. Also it doesn't have a connector at the end of the clamp so that any block including the terminal ones can connect to it.

### Sample

![blocks_sample](https://user-images.githubusercontent.com/33328728/107846685-92b58f80-6e0b-11eb-83c2-32232978c19b.png)

**Comment by walterbender** [comment]
I like the flow design, We could do something interesting with the colors for the connectors.
I am not a fan of the arguments nesting. I think (1) it won't work well for text strings; (2) it won't work well for putting args into args, e.g., a math equation; (3) I don't really buy the argument that it saves space (it seems to exchange vertical space for horizontal space); and (4) it forces some blocks to vary in size vertically.

The whole thing looks a bit too much like Scratch, which is going to suggest it is scratch.

**Reply by meganindya** [reply]
Yes, indeed it looks a lot like scratch. That needs to change. But, don't you think it's more coherent to have arguments side by side with the operator in the middle for binary operations? And, I think using up space horizontally is better than doing it vertically as it is the narrower side.

**Comment by walterbender** [comment]
Maybe something along these lines for nesting argument blocks?
![Screenshot from 2021-02-19 17-17-49](https://user-images.githubusercontent.com/3996641/108567613-913a0a80-72d6-11eb-9041-a0e25cdf3d8c.png)

**Reply by walterbender** [reply]
![Screenshot from 2021-02-19 17-24-11](https://user-images.githubusercontent.com/3996641/108568038-600e0a00-72d7-11eb-950e-10092bd42842.png)

**Reply by meganindya** [reply]
Don't we need some separation between the blocks? How'd a user tell which blocks take tuples and which don't?

**Comment by walterbender** [comment]
![Screenshot from 2021-02-27 08-53-27](https://user-images.githubusercontent.com/3996641/109389185-70c30f00-78d9-11eb-8952-77db48ee1c24.png)

**Comment by pikurasa** [comment]
I'm not sure quite where we are at with this, but I think it's worth exploring using horizontal space more with note blocks. Right now, a stack of note blocks can take up a lot of vertical space.

So:
Note value = `1/4`
Pitch = `Do4`
vs.
Note value =
`
1
`
`
/
`
`
4
`
Pitch =
`
Do
`
`
4
`
